src/external/visualization_integration.cu

```diff
@@ -1,35 +1,76 @@
 #include <iostream>
 #include <fstream>
 #include <string>
+#include <vector>
 
 #include "visualization_integration.cuh"
 #include "../simulation_objects/Particle.cuh"
 
 
-__host__ std::string get_visualization_endpoint()
+__host__ std::pair<std::string, std::string> get_visualization_endpoints()
 {
     std::ifstream f("config/visualization_endpoint.txt", std::ios::in);
-    std::string url;
-    getline(f, url);
-    return url;
+    std::string particles_url, poly_url;
+    getline(f, particles_url);
+    getline(f, poly_url);
+    return {particles_url, poly_url};
 }
 
-__host__ std::string vector_double_to_json_array(const std::vector<double> &v)
+template<class T>
+__host__ std::string to_string_extended(const T &v)
 {
-    if(v.empty())
-        return "[]";
+    return std::to_string(v);
+}
 
-    std::string ans = "[";
-    for(double i : v)
+template<class T>
+__host__ std::string to_string_extended(const std::vector<T> &v)
+{
+    std::string body = "[";
+    for(int i = 0; i < v.size(); ++i)
     {
-        ans += std::to_string(i) + ',';
+        body += to_string_extended(v[i]) + ',';
     }
-    ans.back() = ']';
+    body.back() = ']';
 
-    return ans;
+    return body;
 }
 
-__host__ bool send_particles_to_visualization(const std::string &url, MapNode *nodes, int n_of_nodes)
+__host__ bool send_model_to_visualization(const std::pair<std::string, std::string> &urls, const Polyhedron *polyhedron)
+{
+    std::vector<double> polyhedron_vertices, polyhedron_faces;
+    std::vector<std::vector<double>> poly;
+
+    for(int i = 0; i < polyhedron->get_n_of_faces(); ++i)
+    {
+        for(int j = 0; j < polyhedron->get_faces()[i].get_n_of_vertices(); ++j)
+        {
+            SpacePoint v = polyhedron->get_faces()[i].get_vertices()[j];
```

Comment:
Because the functions `get_n_of_faces`, `get_faces`, `get_n_of_vertices` and `get_vertices` are in a separate source file, I'm not sure the compiler will optimize the calls to these functions to call them only once, so I'd like you to explicitly tell it to do this by introducing local variables which would cache the sizes and `faces`/`vertices` array instead of calling the functions each time you need a value